File: packages/flutter/test/client_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:qr_studio/qr_studio.dart';

// A 4-byte PNG magic header is enough to assert byte plumbing.
final _png = [0x89, 0x50, 0x4e, 0x47];

void main() {
  test('requires a non-empty baseUrl', () {
    expect(() => QrStudioClient(baseUrl: ''), throwsArgumentError);
    expect(() => QrStudioClient(baseUrl: '   '), throwsArgumentError);
  });

  test('normalises baseUrl (trailing slash and /qr suffix) to one /qr endpoint',
      () {
    for (final base in [
      'https://x.dev',
      'https://x.dev/',
      'https://x.dev/qr',
      'https://x.dev/v1/qr',
    ]) {
      final client = QrStudioClient(baseUrl: base);
      expect(client.endpoint.toString(), 'https://x.dev/qr');
      client.close();
    }
  });

  test('buildUri serialises only provided options', () {
    final client = QrStudioClient(baseUrl: 'https://x.dev');
    final uri = client.buildUri(const QrOptions(
      data: 'https://example.com',
      style: QrStyle.brand,
      size: 320,
    ));
    expect(uri.queryParameters['data'], 'https://example.com');
    expect(uri.queryParameters['style'], 'brand');
    expect(uri.queryParameters['size'], '320');
    expect(uri.queryParameters.containsKey('ec'), isFalse);
    client.close();
  });

  test('toParams maps enums to wire values', () {
    final params = const QrOptions(
      data: 'hi',
      format: QrFormat.svg,
      ec: ErrorCorrection.q,
      style: QrStyle.brand,
      verify: false,
      logoPlate: true,
    ).toParams();
    expect(params['format'], 'svg');
    expect(params['ec'], 'Q');
    expect(params['style'], 'brand');
    expect(params['verify'], 'false');
    expect(params['logoPlate'], 'true');
  });

  test('generate POSTs JSON and parses PNG + verified header', () async {
    late http.Request captured;
    final mock = MockClient((req) async {
      captured = req;
      return http.Response.bytes(
        _png,
        200,
        headers: {'content-type': 'image/png', 'x-qr-verified': 'true'},
      );
    });
    final client = QrStudioClient(baseUrl: 'https://x.dev', httpClient: mock);

    final res = await client.generate(const QrOptions(data: 'hi', verify: true));

    expect(captured.method, 'POST');
    expect(captured.url.toString(), 'https://x.dev/qr');
    expect(captured.headers['content-type'], contains('application/json'));
    expect(jsonDecode(captured.body), {'data': 'hi', 'verify': 'true'});

    expect(res.format, QrFormat.png);
    expect(res.verified, isTrue);
    expect(res.bytes.sublist(0, 4), _png);
  });

  test('verified is null when the header is absent', () async {
    final mock = MockClient((req) async {
      return http.Response.bytes(_png, 200, headers: {'content-type': 'image/png'});
    });
    final client = QrStudioClient(baseUrl: 'https://x.dev', httpClient: mock);
    final res = await client.generate(const QrOptions(data: 'hi', verify: false));
    expect(res.verified, isNull);
  });

  test('svg() returns markup', () async {
    const svg = '<svg xmlns="http://www.w3.org/2000/svg"></svg>';
    final mock = MockClient((req) async {
      return http.Response(
        svg,
        200,
        headers: {'content-type': 'image/svg+xml; charset=utf-8'},
      );
    });
    final client = QrStudioClient(baseUrl: 'https://x.dev', httpClient: mock);
    final out = await client.svg(const QrOptions(data: 'hi'));
    expect(out, svg);
  });

  test('non-2xx throws QrStudioException with the API error message', () async {
    final mock = MockClient((req) async {
      return http.Response(
        jsonEncode({'error': '`data` is required'}),
        400,
        headers: {'content-type': 'application/json'},
      );
    });
    final client = QrStudioClient(baseUrl: 'https://x.dev', httpClient: mock);
    expect(
      () => client.generate(const QrOptions(data: 'x')),
      throwsA(isA<QrStudioException>()
          .having((e) => e.statusCode, 'statusCode', 400)
          .having((e) => e.message, 'message', '`data` is required')),
    );
  });

  test('empty data throws before any request', () {
    final client = QrStudioClient(baseUrl: 'https://x.dev');
    expect(() => client.buildUri(const QrOptions(data: '')), throwsArgumentError);
    client.close();
  });
}

File: packages/js/test/client.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { QrStudio, QrStudioError } from '../src/index.ts';

/** A tiny fetch double that records the last call and returns a scripted response. */
function mockFetch(response: {
  status?: number;
  headers?: Record<string, string>;
  body?: Uint8Array | string;
}) {
  const calls: Array<{ url: string; init: RequestInit | undefined }> = [];
  const impl = (async (input: string | URL | Request, init?: RequestInit) => {
    calls.push({ url: String(input), init });
    const body = response.body ?? new Uint8Array([0x89, 0x50, 0x4e, 0x47]); // PNG magic
    const bytes = typeof body === 'string' ? new TextEncoder().encode(body) : body;
    return new Response(bytes as unknown as BodyInit, {
      status: response.status ?? 200,
      headers: response.headers ?? { 'content-type': 'image/png' },
    });
  }) as unknown as typeof fetch;
  return { impl, calls };
}

test('requires a baseUrl', () => {
  assert.throws(() => new QrStudio({ baseUrl: '' } as never), TypeError);
  // @ts-expect-error - intentionally omitting baseUrl
  assert.throws(() => new QrStudio({}), TypeError);
});

test('normalises baseUrl (trailing slash and /qr suffix) to a single /qr endpoint', () => {
  const { impl, calls } = mockFetch({});
  for (const base of ['https://x.dev', 'https://x.dev/', 'https://x.dev/qr', 'https://x.dev/v1/qr']) {
    const qr = new QrStudio({ baseUrl: base, fetch: impl });
    const u = qr.url({ data: 'hi' });
    assert.ok(u.startsWith('https://x.dev/qr?'), `got ${u}`);
  }
  assert.equal(calls.length, 0, 'url() must not perform a request');
});

test('url() serialises only provided options', () => {
  const { impl } = mockFetch({});
  const qr = new QrStudio({ baseUrl: 'https://x.dev', fetch: impl });
  const u = new URL(qr.url({ data: 'https://example.com', style: 'brand', size: 320 }));
  assert.equal(u.searchParams.get('data'), 'https://example.com');
  assert.equal(u.searchParams.get('style'), 'brand');
  assert.equal(u.searchParams.get('size'), '320');
  assert.equal(u.searchParams.get('ec'), null, 'omitted options must not be sent');
});

test('generate() POSTs JSON and parses PNG + verified header', async () => {
  const { impl, calls } = mockFetch({ headers: { 'content-type': 'image/png', 'x-qr-verified': 'true' } });
  const qr = new QrStudio({ baseUrl: 'https://x.dev', fetch: impl });
  const res = await qr.generate({ data: 'hi', verify: true });

  assert.equal(calls.length, 1);
  assert.equal(calls[0].init?.method, 'POST');
  assert.match(String((calls[0].init?.headers as Record<string, string>)['Content-Type']), /application\/json/);
  assert.deepEqual(JSON.parse(String(calls[0].init?.body)), { data: 'hi', verify: 'true' });

  assert.equal(res.format, 'png');
  assert.equal(res.verified, true);
  assert.deepEqual([...res.bytes.subarray(0, 4)], [0x89, 0x50, 0x4e, 0x47]);
});

test('verified is null when the header is absent', async () => {
  const { impl } = mockFetch({ headers: { 'content-type': 'image/png' } });
  const qr = new QrStudio({ baseUrl: 'https://x.dev', fetch: impl });
  const res = await qr.generate({ data: 'hi', verify: false });
  assert.equal(res.verified, null);
});

test('svg() returns markup and sets format', async () => {
  const svg = '<svg xmlns="http://www.w3.org/2000/svg"></svg>';
  const { impl } = mockFetch({ headers: { 'content-type': 'image/svg+xml; charset=utf-8' }, body: svg });
  const qr = new QrStudio({ baseUrl: 'https://x.dev', fetch: impl });
  const out = await qr.svg({ data: 'hi' });
  assert.equal(out, svg);
});

test('dataUrl() base64-encodes the bytes with the right mime', async () => {
  const png = new Uint8Array([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a]);
  const { impl } = mockFetch({ headers: { 'content-type': 'image/png' }, body: png });
  const qr = new QrStudio({ baseUrl: 'https://x.dev', fetch: impl });
  const url = await qr.dataUrl({ data: 'hi' });
  assert.ok(url.startsWith('data:image/png;base64,'));
  const b64 = url.slice('data:image/png;base64,'.length);
  assert.deepEqual([...Buffer.from(b64, 'base64')], [...png]);
});

test('non-2xx throws QrStudioError with the API error message', async () => {
  const { impl } = mockFetch({
    status: 400,
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ error: '`data` is required' }),
  });
  const qr = new QrStudio({ baseUrl: 'https://x.dev', fetch: impl });
  await assert.rejects(
    () => qr.generate({ data: 'x' }),
    (err: unknown) => {
      assert.ok(err instanceof QrStudioError);
      assert.equal(err.status, 400);
      assert.equal(err.message, '`data` is required');
      return true;
    },
  );
});

test('requires data', () => {
  const { impl } = mockFetch({});
  const qr = new QrStudio({ baseUrl: 'https://x.dev', fetch: impl });
  assert.throws(() => qr.url({ data: '' }), TypeError);
});
